salt/transport/rabbitmq.py

```diff
@@ -0,0 +1,1243 @@
+    def __del__(self):
+        self.close()
+
+    # pylint: enable=W1701
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+    # TODO: this is the time to see if we are connected, maybe use the req channel to guess?
+    @salt.ext.tornado.gen.coroutine
+    def connect(self):
+        """
+        Connects minion to master.
+        TODO: anything else that needs to be done for broker-based connections?
+        :return:
+        """
+
+        if not self.auth.authenticated:
+            yield self.auth.authenticate()
+
+        log.debug(
+            "Minion already authenticated with master. Nothing else to do for broker-based transport.")
+
+    @salt.ext.tornado.gen.coroutine
+    def _decode_messages(self, messages):
+        """
+        Take the rmq messages, decrypt/decode them into a payload
+
+        :param list messages: A list of messages to be decoded
+        """
+        messages = [messages] # TODO: FIXME - figure out why this does not match zeromq payload packing
+        messages_len = len(messages)
+        # if it was one message, then its old style
+        if messages_len == 1:
+            payload = self.serial.loads(messages[0])
+            payload = self.serial.loads(payload["payload"])
+        # 2 includes a header which says who should do it
+        elif messages_len == 2:
+            message_target = salt.utils.stringutils.to_str(messages[0])
+            if (
+                self.opts.get("__role") != "syndic"
+                and message_target not in ("broadcast", self.hexid)
+            ) or (
+                self.opts.get("__role") == "syndic"
+                and message_target not in ("broadcast", "syndic")
+            ):
+                log.debug("Publish received for not this minion: %s", message_target)
+                raise salt.ext.tornado.gen.Return(None)
+            payload = self.serial.loads(messages[1])
+        else:
+            raise Exception(
+                (
+                    "Invalid number of messages ({}) in rabbitmq pub"
+                    "message from master"
+                ).format(len(messages_len))
+            )
+        # Yield control back to the caller. When the payload has been decoded, assign
+        # the decoded payload to 'ret' and resume operation
+        ret = yield self._decode_payload(payload)
+        raise salt.ext.tornado.gen.Return(ret)
+
+    def on_recv(self, callback):
+        """
+        Register an on_recv callback
+        """
+        if callback is None:
+            self._rmq_non_blocking_connection_wrapper.register_message_callback(callback=None)
+
+        @salt.ext.tornado.gen.coroutine
+        def wrap_callback(messages):
+            payload = yield self._decode_messages(messages)
+            if payload is not None:
+                callback(payload)
+
+        self._rmq_non_blocking_connection_wrapper.register_message_callback(callback=wrap_callback)
```

Comment:
**note** I see a lot of tornado stuff, and I don't recall seeing a lot of tornado-based testing. I don't know if that's covered by the existing tests, but especially for things like authentication we should have some tests around that :+1: